hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java

```diff
@@ -143,8 +142,12 @@ public static void logSqlException(Throwable t, Supplier<String> message, String
 	 * }
 	 * </pre>
 	 */
-	public static <T> CompletionStage<Void> loop(T[] array, Function<T,CompletionStage<?>> consumer) {
-		return loop( Arrays.stream(array), consumer );
+	public static <T> CompletionStage<Void> loop(T[] array, Function<T, CompletionStage<?>> consumer) {
+		AtomicInteger index = new AtomicInteger();
+		return AsyncTrampoline.asyncWhile( () -> {
+			int i = index.getAndIncrement();
+			return consumer.apply( array[i] ).thenApply( v -> i < array.length - 1 );
+		} );
```

Comment:
Are we really certain that creating an `AtomicInteger` is less expensive than creating a stream? I mean, it certainly sounds _possible_, but it's not _obviously_ true.

Reply:
Remember: "creating a stream" in this case just means making a little wrapper for an array.